### ISSUE TITLE
Possible race if PR is reconciled before Pipeline

### ISSUE THREAD

**Issue by stuartwdouglas** [tektoncd/pipeline#5601]
When creating a PipelineRun and Pipeline in quick succession you can get the following error:
```
    types.go:25: time: 2022-10-04 21:36:41.377386386 +0000 UTC m=+47.592144259: non-normal event reason Failed about obj PipelineRun:sample-component-build-bh8mx message Pipeline jvm-build-service-test-namespace-mzhhd/sample-component-build can't be Run; it has an invalid spec: "" type does not match default value's type: "string": finally.params[app-path].default.type, finally.params[app-path].type, finally.params[context].default.type, finally.params[context].type, finally.params[revision].default.type, finally.params[revision].type, tasks.params[app-path].default.type, tasks.params[app-path].type, tasks.params[context].default.type, tasks.params[context].type, tasks.params[revision].default.type, tasks.params[revision].type
```
e.g.  as at https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/290/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-e2e/1577408946094739456

I had a look at the code and I am pretty sure this happens if the PR is reconciled first, and the pipeline has parameters that have a default but no explicit type set. Normally the first reconcile of the Pipeline would set this type, however if the PR is reconciled before the pipeline this won't happen (note that this is just my theory, I have no real knowledge of the code base, so I may be wrong). 

# Steps to Reproduce the Problem

It is racy, so it is hard to reproduce consistently. I have seen it multiple times, but fairly infrequently. It usually happens if you use a script to both install tekton, the pipeline and the pipelinerun straight after each other.

Generally once it happens you need to delete the pipeline and start again.

# Additional Info

- Kubernetes version:

Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v0.21.0-beta.1", GitCommit:"0d10c3f72592addce965b9bb34992eb6fc283a3b", GitTreeState:"clean", BuildDate:"2021-08-31T22:03:33Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.5+012e945", GitCommit:"3c28e7a79b58e78b4c1dc1ab7e5f6c6c2d3aedd3", GitTreeState:"clean", BuildDate:"2022-07-13T08:38:41Z", GoVersion:"go1.17.12", Compiler:"gc", Platform:"linux/amd64"}

  ```
  (paste your output here)
  ```

- Tekton Pipeline version:

Client version: 0.26.0
Chains version: v0.9.0
Pipeline version: v0.37.4
Triggers version: v0.20.2
Operator version: v0.60.1

(note that this is from my personal openshift, I am not sure what version is run on openshift CI above).

**Comment by vdemeester** [comment]
Hey @stuartwdouglas, thanks for the issue. One thing though, `Pipeline` are not "reconciled" at all, only `PipelineRun` and `TaskRun` are — but `Pipeline` and `Task` (and other type) do have a mutating admission controller though. 
Any specifics on the Pipeline used ? (like using pipelineref with oci bundles, or something like that).
Also, where does this errors pops out ? in the controller logs or in the webhook logs ?

If it sometimes works, and sometimes it doesn't, and given the names in the logs, one thing that "could" happen here is, that some object are created *before* the admission controller registered itself, making it bypass any validation or "defaulting". Essentially this would mean some object are created before the pipeline instance is fully ready to work.

**Comment by stuartwdouglas** [comment]
Ah, looking again I seem to have misunderstood what what going on here. 

Looking at the actual code I don't actually understand what is going on, as far as I can see the message is generated here: https://github.com/tektoncd/pipeline/blob/a11c38585b924d9b07b93bc38524889c7413ce1d/pkg/apis/pipeline/v1beta1/task_validation.go#L305

The message `"" type does not match default value's type: "string"` implies that p.Type is the empty string, however the code above that should have caught that problem. 

Looking at the code I have no idea how that message could be generated, but it clearly is as shown in the logs.

**Comment by vdemeester** [comment]
So we looked at this offline with @stuartwdouglas, and this is indeed a "timing" issue. 
Because the webhook registers itself, there is a small (but possibly long) period of time, where the CRDs are available *but* the webhook is not registered (or has no rules). This is not really a problem in a "normal" usage, but for things like automated e2e tests, it could be a problem, as one could create some object before the admission controller are ready, making it bypass validation or mutation (defaulting, conversion).

Initially, I thought we could check for the webhook pod readiness, but that doesn't work either. As of today (see https://github.com/tektoncd/pipeline/blob/main/cmd/webhook/main.go#L219), we mark the webhook pod as ready as soon as it start. Ideally we should mark it ready *when* we registered. This might need some knative/pkg changes though.

/cc @imjasonh @abayer
